book_lovers/my_books/api_views.py

```diff
@@ -21,47 +13,80 @@ def has_permission(self, request, view):
 
     def has_object_permission(self, request, view, obj):
         if request.user.is_authenticated():
-            if (obj.isPublished) or ((request.user.profile.publisher == obj.publisher) or (obj in request.user.owned_books.all())):
-                return True
-            else:
-                return False
+           
+            if request.method in permissions.SAFE_METHODS:
+ 
+                if (obj.is_public()) \
+                        or request.user.is_staff\
+                        or ((request.user.profile.publisher == obj.publisher)
+                            or (obj in request.user.uploaded_books.all())
+                            or (obj in request.user.authored_books.all())):
+                    return True
+                else:
+                    return False
+            else:  # editing permissions
+                if obj.is_public():
+                    if request.user.is_staff:
+                        return True
+                    
+                else:
+                    if request.user.is_staff\
+                                    or ((request.user.profile.publisher == obj.publisher)
+                                        or (obj in request.user.uploaded_books.all())
+                                        or (obj in request.user.authored_books.all())):
+                        return True
+            
         return False
 
+    
 class BookViewSet(viewsets.ModelViewSet):
-    queryset = Book.objects.all()
-    serializer_class =BookSerializer
-    filter_fields = ('title',)
-
-    def perform_create(self, serializer):
-        serializer.save(owner=self.request.user)  
-    #permission_classes = (permissions.IsAuthenticatedOrReadOnly,)
+    def get_queryset(self):
+        if self.request.user.is_staff:
+            return Book.objects.all()
+        else:
+            return Book.objects.filter(Q(isVerified=True, isPublished=True) | Q(uploader=self.request.user)
+                                       | Q(author=self.request.user) | Q(publisher=self.request.user.profile.publisher))
+
+    queryset = Book.objects.get_queryset()
+    
+    filter_fields = ('isPublished', 'isVerified')
     permission_classes = (BookViewPermission,)
 
+    def get_serializer_class(self):
+        # if the user is admin, use the BookAdminSerializer. For any other user, the base serializer
+        if self.request.user.is_staff:
+            return BookAdminSerializer
+        else:  # non-admin user
+            return BookSerializer
 
-    def patch(self, request):
-        data = request.data
-        testmodel = (data.get('id'))
-        serializer = BookSerializer(data=request.data, partial=True) # set partial=True to update a data partially
-        if serializer.is_valid():
+    def perform_create(self, serializer):
+        data = self.request.data
+        post_type = data.__getitem__('type')
```

Comment:
why are you calling __getitem__ like that as opposed to something like ['type'] ? or .get( ) ?